contentcuration/contentcuration/frontend/shared/data/serverSync.js

```diff
@@ -212,82 +207,92 @@ async function syncChanges(syncAllChanges) {
-      const changesToSync = await syncableChanges.toArray();
-      // By the time we get here, our changesToSync Array should
-      // have every change we want to sync to the server, so we
-      // can now trim it down to only what is needed to transmit over the wire.
-      // TODO: remove moves when a delete change is present for an object,
-      // because a delete will wipe out the move.
-      const changes = changesToSync.map(trimChangeForSync).filter(Boolean);
-      // Create a promise for the sync - if there is nothing to sync just resolve immediately,
-      // in order to still call our change cleanup code.
-      if (changes.length) {
-        requestPayload.changes = changes;
+      const requestPayload = {
+        changes: [],
+        channel_revs,
+        user_rev: user.user_rev || 0,
+        unapplied_revs: unAppliedChanges.map(c => c.server_rev).filter(Boolean),
+      };
+
+      // Snapshot which revs we are syncing, so that we can
+      // removes them from the changeRevs array after the sync
+      const revsToSync = [];
+      if (syncAllChanges) {
+        const unsyncedRevs = await db[CHANGES_TABLE].filter(c => !c.synced).primaryKeys();
+        revsToSync.push(...unsyncedRevs);
+      } else {
+        revsToSync.push(...changeRevs);
+      }
+      if (revsToSync.length) {
+        const syncableChanges = db[CHANGES_TABLE].where('rev')
+          .anyOf(revsToSync)
+          .filter(c => !c.synced);
+        const changesToSync = await syncableChanges.toArray();
+        // By the time we get here, our changesToSync Array should
+        // have every change we want to sync to the server, so we
+        // can now trim it down to only what is needed to transmit over the wire.
+        // TODO: remove moves when a delete change is present for an object,
+        // because a delete will wipe out the move.
+        const changes = changesToSync.map(trimChangeForSync).filter(Boolean);
+        // Create a promise for the sync - if there is nothing to sync just resolve immediately,
+        // in order to still call our change cleanup code.
+        if (changes.length) {
+          requestPayload.changes = changes;
+        }
+      }
+      // The response from the sync endpoint has the format:
+      // {
+      //   "disallowed": [],
+      //   "allowed": [],
+      //   "changes": [],
+      //   "errors": [],
+      //   "successes": [],
+      // }
+      const response = await client.post(urls['sync'](), requestPayload);
+      // Clear out this many changes from changeRevs array, since we have now synced them.
+      changeRevs.splice(0, revsToSync.length);
+      await Promise.all([
+        handleDisallowed(response),
+        handleAllowed(response),
+        handleErrors(response),
+        handleSuccesses(response),
+        handleMaxRevs(response, user.id),
+        handleTasks(response),
+      ]);
+      // Return the array of returned changes
+      return get(response, ['data', 'changes'], []);
```

Comment:
We would need to do it before we apply `handleMaxRevs` but one thing I had been thinking about was whether we should filter out changes here which the max revision values exceed before we return them for processing.

The locks should take care of most duplicate changes, but it still seems plausible that we could have duplicate user changes here, so maybe before we do `handleMaxRevs` we get the currently recorded max user rev and filter out any changes with the `user_id` set to `user.id` that have a server_rev lower than that max rev?

Reply:
Seems like we'd need the returned change application and maxRevs handling all in one lock then?

Reply:
Hrm, yeah - maybe we should be incrementing the max revs as we apply each change?

Reply:
Yes, but if I understand correctly, then that would also mean that we would have to check all changes, not just user-related changes, since the sync sends the current max rev to the server and the server could return duplicate changes if another tab's sync was waiting for the current sync and the sync continues to execute in a separate lock?

Reply:
Yes - I think that's why I was suggesting in the original comment to just record the current max revs before we set them from the response, and use that to filter the changes before we return them for application. So the sequence would be:

1. Do the sync request.
2. Read the current max revs before we set them from the response
3. Apply max revs etc.
4. Return the returned changes, filtered by the max revs that we read before applying max revs from the response.
I think that should minimize the chance of duplicate changes being applied?

Reply:
Updated to keep both the sync and change application in the same lock for now